File: python/h5PYRe/params.py
"""Read/write the /params group (see ../FORMAT.md)."""

import os

import h5py
import numpy as np
import yaml

from .io import put

_SCALAR_TYPES = (bool, int, float, str)


class _QuotedStrDumper(yaml.SafeDumper):
    """SafeDumper that emits every string scalar single-quoted.

    Necessary for cross-language fidelity: PyYAML's own scalar resolver does
    not treat single-letter tokens like ``n``/``y`` as booleans, so it would
    emit them bare, but R's ``yaml`` package (full YAML 1.1) reads a bare
    ``n`` as ``FALSE``. Quoting every string (keys included) keeps param names
    and values type-stable when the file is read back in either language.
    """


def _quoted_str(dumper, value):
    return dumper.represent_scalar("tag:yaml.org,2002:str", value, style="'")


_QuotedStrDumper.add_representer(str, _quoted_str)


def _normalize(value):
    """Coerce a ``read_params`` value to plain Python for YAML serialization."""
    if isinstance(value, np.ndarray):
        return value.tolist()
    if isinstance(value, np.generic):
        return value.item()
    if isinstance(value, list):
        return [v.item() if isinstance(v, np.generic) else v for v in value]
    return value


def _check_param_value(name, value):
    """Reject YAML values that ``/params`` (1-D vectors of scalars) can't hold."""
    if isinstance(value, _SCALAR_TYPES):
        return
    if isinstance(value, list) and all(isinstance(v, _SCALAR_TYPES) for v in value):
        return
    raise ValueError(
        f"import_params(): param {name!r} must be a scalar or a flat list of "
        f"scalars; nested or complex values are not supported by /params"
    )


def write_params(path, params):
    """Write a dict of parameters to ``/params``.

    Creates the file if it does not exist, otherwise appends/overwrites.
    """
    mode = "r+" if os.path.exists(path) else "w"
    with h5py.File(path, mode) as f:
        g = f.require_group("params")
        for name, value in params.items():
            put(g, name, value)
    return path


def read_params(path):
    """Read ``/params`` as a dict.

    Length-1 datasets are unwrapped to Python scalars; longer ones are returned
    as numpy arrays (numeric) or lists of str (text). Empty dict if no
    ``/params`` group.
    """
    out = {}
    with h5py.File(path, "r") as f:
        if "params" not in f:
            return out
        g = f["params"]
        for name in g:
            vals = np.ravel(g[name][()])
            if vals.dtype.kind in ("S", "O"):
                vals = [v.decode() if isinstance(v, bytes) else v for v in vals]
                out[name] = vals[0] if len(vals) == 1 else vals
            else:
                out[name] = vals[0].item() if len(vals) == 1 else vals
    return out


def export_params(path, out_file, params=None):
    """Export ``/params`` to a YAML file at ``out_file``.

    By default exports every parameter; pass ``params`` (a list of names) to
    export only a subset. Strings are emitted single-quoted so the file reads
    back with identical types in both Python and R (see ``_QuotedStrDumper``).

    YAML is a lossy-unless-quoted *external* view of the params, not part of
    the on-disk format. ``import_params`` reverses this export faithfully.
    Returns ``out_file``.
    """
    values = read_params(path)
    if params is not None:
        missing = [p for p in params if p not in values]
        if missing:
            raise KeyError(f"export_params(): no such param(s): {missing}")
        values = {p: values[p] for p in params}
    values = {name: _normalize(v) for name, v in values.items()}

    parent = os.path.dirname(out_file)
    if parent:
        os.makedirs(parent, exist_ok=True)
    with open(out_file, "w", encoding="utf-8") as fh:
        yaml.dump(values, fh, Dumper=_QuotedStrDumper, default_flow_style=False,
                  sort_keys=False, allow_unicode=True)
    return out_file


def import_params(path, yaml_file, params=None, replace=False):
    """Write params from a YAML file into ``/params`` of ``path``.

    Reverses :func:`export_params`. Reads ``yaml_file`` (a mapping of param
    name -> scalar or flat list of scalars), optionally restricts to
    ``params`` (a list of names), and writes them via ``write_params`` —
    merging into any existing ``/params`` by default (per-key overwrite). Pass
    ``replace=True`` to clear ``/params`` first. Creates ``path`` if needed.

    Returns the dict of params written.
    """
    with open(yaml_file, encoding="utf-8") as fh:
        loaded = yaml.safe_load(fh)
    if loaded is None:
        loaded = {}
    if not isinstance(loaded, dict):
        raise ValueError(
            f"import_params(): {yaml_file!r} must contain a top-level mapping "
            f"of param name -> value, got {type(loaded).__name__}"
        )
    for name, value in loaded.items():
        _check_param_value(name, value)

    if params is not None:
        missing = [p for p in params if p not in loaded]
        if missing:
            raise KeyError(f"import_params(): no such param(s) in file: {missing}")
        loaded = {p: loaded[p] for p in params}

    if replace and os.path.exists(path):
        with h5py.File(path, "r+") as f:
            if "params" in f:
                del f["params"]

    write_params(path, loaded)
    return loaded

File: r/R/params.R
#' Write parameters to the `/params` group of an HDF5 file
#'
#' Creates the file if needed, otherwise appends/overwrites into it. Each
#' element of `params` is stored as a dataset under `/params/`.
#'
#' @param path Output file path.
#' @param params A named list of vectors (scalars stored as length-1).
#' @return `path`, invisibly.
#' @export
write_params <- function(path, params) {
  stopifnot(is.list(params), !is.null(names(params)))
  h5 <- hdf5r::H5File$new(path, if (file.exists(path)) "r+" else "w")
  on.exit(h5$close_all(), add = TRUE)

  g <- if (h5$exists("params")) h5[["params"]] else h5$create_group("params")
  for (nm in names(params)) h5_put(g, nm, params[[nm]])
  invisible(path)
}

#' Read the `/params` group as a named list
#'
#' @param path Input file path.
#' @return A named list of parameter vectors (empty list if no `/params`).
#' @export
read_params <- function(path) {
  h5 <- hdf5r::H5File$new(path, "r")
  on.exit(h5$close_all(), add = TRUE)
  if (!h5$exists("params")) return(list())

  g   <- h5[["params"]]
  nms <- names(g)
  stats::setNames(lapply(nms, function(nm) g[[nm]]$read()), nms)
}

#' Export the `/params` group to a YAML file
#'
#' By default exports every parameter; pass `params` (a character vector of
#' names) to export only a subset. YAML is a lossy-unless-quoted *external*
#' view of the params (not part of the on-disk format); `import_params()`
#' reverses this export faithfully. R's `yaml` emitter already quotes the
#' YAML-1.1 ambiguous tokens (e.g. `no`, `n`, `on`), so files it writes read
#' back with identical types in both R and Python.
#'
#' @param path Input h5PYRe file.
#' @param out_file Output YAML file path.
#' @param params Optional character vector of parameter names to export.
#' @return `out_file`, invisibly.
#' @export
export_params <- function(path, out_file, params = NULL) {
  values <- read_params(path)
  if (!is.null(params)) {
    missing <- setdiff(params, names(values))
    if (length(missing))
      stop("export_params(): no such param(s): ", paste(missing, collapse = ", "))
    values <- values[params]
  }
  parent <- dirname(out_file)
  if (nzchar(parent) && !dir.exists(parent))
    dir.create(parent, recursive = TRUE, showWarnings = FALSE)
  yaml::write_yaml(values, out_file)
  invisible(out_file)
}

#' Write params from a YAML file into the `/params` group
#'
#' Reverses [export_params()]. Reads `yaml_file` (a mapping of param name to a
#' scalar or flat vector of scalars), optionally restricts to `params`, and
#' writes them via [write_params()] — merging into any existing `/params` by
#' default (per-key overwrite). Pass `replace = TRUE` to clear `/params`
#' first. Creates `path` if needed.
#'
#' @param path Output h5PYRe file.
#' @param yaml_file Input YAML file path.
#' @param params Optional character vector of parameter names to import.
#' @param replace If `TRUE`, delete the existing `/params` group before writing.
#' @return The named list of params written, invisibly.
#' @export
import_params <- function(path, yaml_file, params = NULL, replace = FALSE) {
  loaded <- yaml::read_yaml(yaml_file)
  if (is.null(loaded)) loaded <- list()
  if (!is.list(loaded) || is.null(names(loaded)) || any(!nzchar(names(loaded))))
    stop("import_params(): '", yaml_file,
         "' must contain a top-level mapping of param name -> value")
  for (nm in names(loaded)) {
    v <- loaded[[nm]]
    if (is.list(v) || !is.atomic(v))
      stop("import_params(): param '", nm, "' must be a scalar or a flat ",
           "vector of scalars; nested or complex values are not supported")
  }
  if (!is.null(params)) {
    missing <- setdiff(params, names(loaded))
    if (length(missing))
      stop("import_params(): no such param(s) in file: ",
           paste(missing, collapse = ", "))
    loaded <- loaded[params]
  }
  if (replace && file.exists(path)) {
    h5 <- hdf5r::H5File$new(path, "r+")
    if (h5$exists("params")) h5$link_delete("params")
    h5$close_all()
  }
  write_params(path, loaded)
  invisible(loaded)
}
